File: tests/abi/parameter_helpers_layout.cpp
#include "pluginterfaces/vst/ivstparameterfunctionname.h"
#include "pluginterfaces/vst/ivstplugview.h"
#include "pluginterfaces/vst/ivstremapparamid.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("FunctionNameType.kCompGainReduction %s\n", Steinberg::Vst::FunctionNameType::kCompGainReduction);
	std::printf ("FunctionNameType.kCompGainReductionMax %s\n", Steinberg::Vst::FunctionNameType::kCompGainReductionMax);
	std::printf ("FunctionNameType.kCompGainReductionPeakHold %s\n", Steinberg::Vst::FunctionNameType::kCompGainReductionPeakHold);
	std::printf ("FunctionNameType.kCompResetGainReductionMax %s\n", Steinberg::Vst::FunctionNameType::kCompResetGainReductionMax);
	std::printf ("FunctionNameType.kLowLatencyMode %s\n", Steinberg::Vst::FunctionNameType::kLowLatencyMode);
	std::printf ("FunctionNameType.kDryWetMix %s\n", Steinberg::Vst::FunctionNameType::kDryWetMix);
	std::printf ("FunctionNameType.kRandomize %s\n", Steinberg::Vst::FunctionNameType::kRandomize);
	std::printf ("FunctionNameType.kPanPosCenterX %s\n", Steinberg::Vst::FunctionNameType::kPanPosCenterX);
	std::printf ("FunctionNameType.kPanPosCenterY %s\n", Steinberg::Vst::FunctionNameType::kPanPosCenterY);
	std::printf ("FunctionNameType.kPanPosCenterZ %s\n", Steinberg::Vst::FunctionNameType::kPanPosCenterZ);

	PRINT_TYPE (IParameterFunctionName);
	PRINT_TYPE (IParameterFinder);
	PRINT_TYPE (IRemapParamID);

	print_iid ("IParameterFunctionName", Steinberg::Vst::IParameterFunctionName_iid);
	print_iid ("IParameterFinder", Steinberg::Vst::IParameterFinder_iid);
	print_iid ("IRemapParamID", Steinberg::Vst::IRemapParamID_iid);
	return 0;
}

File: tools/parameter_helpers_layout.zig
const std = @import("std");
const function_name = @import("vst3-zig").pluginterfaces.vst.ivstparameterfunctionname;
const parameter_finder = @import("vst3-zig").pluginterfaces.vst.ivstplugview;
const remap = @import("vst3-zig").pluginterfaces.vst.ivstremapparamid;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("FunctionNameType.kCompGainReduction {s}\n", .{std.mem.span(function_name.FunctionNameType.kCompGainReduction)});
    try stdout.print("FunctionNameType.kCompGainReductionMax {s}\n", .{std.mem.span(function_name.FunctionNameType.kCompGainReductionMax)});
    try stdout.print("FunctionNameType.kCompGainReductionPeakHold {s}\n", .{std.mem.span(function_name.FunctionNameType.kCompGainReductionPeakHold)});
    try stdout.print("FunctionNameType.kCompResetGainReductionMax {s}\n", .{std.mem.span(function_name.FunctionNameType.kCompResetGainReductionMax)});
    try stdout.print("FunctionNameType.kLowLatencyMode {s}\n", .{std.mem.span(function_name.FunctionNameType.kLowLatencyMode)});
    try stdout.print("FunctionNameType.kDryWetMix {s}\n", .{std.mem.span(function_name.FunctionNameType.kDryWetMix)});
    try stdout.print("FunctionNameType.kRandomize {s}\n", .{std.mem.span(function_name.FunctionNameType.kRandomize)});
    try stdout.print("FunctionNameType.kPanPosCenterX {s}\n", .{std.mem.span(function_name.FunctionNameType.kPanPosCenterX)});
    try stdout.print("FunctionNameType.kPanPosCenterY {s}\n", .{std.mem.span(function_name.FunctionNameType.kPanPosCenterY)});
    try stdout.print("FunctionNameType.kPanPosCenterZ {s}\n", .{std.mem.span(function_name.FunctionNameType.kPanPosCenterZ)});

    try printTuid(stdout, "IParameterFunctionName", function_name.iparameter_function_name_iid);
    try printTuid(stdout, "IParameterFinder", parameter_finder.iparameter_finder_iid);
    try printTuid(stdout, "IRemapParamID", remap.iremap_param_id_iid);
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
